analyzers/src/SonarAnalyzer.CSharp/Common/Walkers/CatchLoggingInvocationWalker.cs

```diff
@@ -0,0 +1,119 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using static Roslyn.Utilities.SonarAnalyzer.Shared.LoggingFrameworkMethods;
+
+namespace SonarAnalyzer.Common.Walkers;
+
+// The walker is used to:
+// - visit the catch clause (all the nested catch clauses are skipped; they will be visited independently)
+// - save the declared exception
+// - find all the logging invocations and check if the exception is logged
+// - if the exception is logged, it will stop looking for the other invocations and set IsExceptionLogged to true
+// - if the exception is not logged, it will visit all the invocations
+public sealed class CatchLoggingInvocationWalker : SafeCSharpSyntaxWalker
+{
+    private readonly SemanticModel model;
+    private bool isFirstCatchClauseVisited;
+    private bool hasWhenFilterWithDeclarations;
+
+    public ISymbol CaughtException { get; private set; }
+    public bool IsExceptionLogged { get; private set; }
+    public InvocationExpressionSyntax LoggingInvocationWithException { get; private set; }
+    public ThrowStatementSyntax ThrowStatementSyntax { get; private set; }
+    public ImmutableArray<InvocationExpressionSyntax> LoggingInvocationsWithoutException { get; private set; } = ImmutableArray<InvocationExpressionSyntax>.Empty;
+
+    private static readonly ImmutableArray<LoggingInvocationDescriptor> LoggingInvocationIdentifiers = ImmutableArray.Create(
+        new LoggingInvocationDescriptor(MicrosoftExtensionsLogging, KnownType.Microsoft_Extensions_Logging_LoggerExtensions, false),
+        new LoggingInvocationDescriptor(CastleCoreOrCommonCore, KnownType.Castle_Core_Logging_ILogger, true),
+        new LoggingInvocationDescriptor(CastleCoreOrCommonCore, KnownType.Common_Logging_ILog, true),
+        new LoggingInvocationDescriptor(Log4NetILog, KnownType.log4net_ILog, true),
+        new LoggingInvocationDescriptor(Log4NetILogExtensions, KnownType.log4net_Util_ILogExtensions, false),
+        new LoggingInvocationDescriptor(NLogLoggingMethods, KnownType.NLog_ILogger, true),
+        new LoggingInvocationDescriptor(NLogLoggingMethods, KnownType.NLog_ILoggerExtensions, false),
+        new LoggingInvocationDescriptor(NLogILoggerBase, KnownType.NLog_ILoggerBase, true),
+        new LoggingInvocationDescriptor(Serilog, KnownType.Serilog_ILogger, true),
+        new LoggingInvocationDescriptor(Serilog, KnownType.Serilog_Log, false));
+
+    public CatchLoggingInvocationWalker(SemanticModel model)
```

Comment:
I think this walker should override both of the lambda boys to do nothing:

```csharp
    public override void VisitParenthesizedLambdaExpression(ParenthesizedLambdaExpressionSyntax node) { }
    public override void VisitSimpleLambdaExpression(SimpleLambdaExpressionSyntax node) { }
```

Because of cases like these:

```csharp
    public void LambdaTricks(ILogger logger)
    {
        try
        { }
        catch (Exception hotPotato) // Noncompliant FP
        {
            Action x = () => { throw hotPotato; }; // Secondary - Annoying
            logger.LogError(hotPotato, "Message"); // Secondary
        }

        try
        { }
        catch (Exception hotPotato) // Noncompliant FP
        {
            Action x = () => { logger.LogError(hotPotato, "Message"); }; // Secondary - Annoying
            throw hotPotato; // Secondary
        }
    }
```

You will be having some FNs with this logic (for example when the `Action` is actually called, and not just declared), but FNs are better than FPs.

Reply:
Good point!